### DISCUSSION TITLE
Best Way to Get Boundary Condition of Family Node (Import of Pointwise Grid)?

### DISCUSSION THREAD

**Discussion by MamboJambo42** [CGNS/CGNS#308] [CGNStalk]
We want to use CGNS grids for our block-structured CFD solver. Our primary software for meshing is Pointwise. It seems that Pointwise doesn't use the standard BC functions of CGNS to write the BC data, but uses Family nodes instead. So calling `cg_boco_info` on a BC only returns `FamilySpecified`.

My question is: Given such a CGNS file, what is the correct/best way to get the actual BC type when looping over all BCs of a zone? 

The way I got it working feels very awkward and might be caused by the fact that I'm new to the CGNS library: 
1. call `cg_boco_info` to find out that the BC type is  `FamilySpecified`
2. `cs_goto` the respective `BC_t` node
2. read the corresponding family name with `cg_famname_read`
3. get the number of families with `cg_nfamilies`
4. Loop over all families and call `cg_family_read` to compare the name of each family to the family name of the actual BC, and then call `cg_fambc_read` to get the BC type of the matching family 

Of course, I could read and store the number of families and the associated `FamilyBC_t` once at the beginning, but I would still have to compare the Family names. It feels like there should be a more elegant way. Am I missing something here?

**Comment by MicK7** [comment]
After looking into the cgns utility "cgns_to_aflr3" provided with the library, it seems you are using the correct way.
You should really read all Family_t node once at the beginning, to prevent doing a lot of goto or gopath.
If you use a dedicated structure to store the families you can optimize the string comparison.

Pointwise is providing the current best practice of CGNS standard, aka use Family to define Boundary Conditions.

What do you think is missing ? Or how do you expect the mll to behave ?

**Comment by MamboJambo42** [comment]
MicK7, thank you for your comments! 

I was wondering if it is possible to skip the whole Family name comparison. If `cg_boco_info` or `cg_famname_read` would also return the Family Index of the associated Family of the BC, I could directly call `cg_fambc_read` to retrieve the BC type I'm looking for. Obtaining the Family Name and then filtering for the Family Index I actually need feels somewhat tedious to me.

**Reply by MicK7** [reply]
Recent CPEX have changed the status of families and now it is possible to have hierarchies. Thus it is not posssible to return a Family index in the MLL without breaking the standard. 
If you are using C++ you can store the family information in a unordered_map and when you get the Family in the BC it will be very easy to retrieve the true BC type.

I agree that the MLL is quite old and maybe at some point it will need a rewrite to make parsing easier.

**Comment by MamboJambo42** [comment]
Ok, I understand. Thank you for the insight.
